src/category_theory/sites/pretopology.lean

```diff
@@ -0,0 +1,274 @@
+               (Ti : Π ⦃Y⦄ (f : Y ⟶ X), S f → arrows_with_codomain Y), S ∈ coverings X →
+               (∀ ⦃Y⦄ f (H : S f), Ti f H ∈ coverings Y) → S.bind Ti ∈ coverings X)
+
+namespace pretopology
+
+instance : has_coe_to_fun (pretopology C) :=
+⟨_, λ J, J.coverings⟩
+
+instance : partial_order (pretopology C) :=
+{ le := λ K₁ K₂, (K₁ : Π (X : C), set _) ≤ K₂,
+  le_refl := λ K, le_refl _,
+  le_trans := λ K₁ K₂ K₃ h₁₂ h₂₃, le_trans h₁₂ h₂₃,
+  le_antisymm := λ K₁ K₂ h₁₂ h₂₁, pretopology.ext _ _ (le_antisymm h₁₂ h₂₁) }
+
+instance : order_top (pretopology C) :=
+{ top :=
+  { coverings := λ _, set.univ,
+    has_isos := λ _ _ _ _, set.mem_univ _,
+    pullbacks := λ _ _ _ _ _, set.mem_univ _,
+    transitive := λ _ _ _ _ _, set.mem_univ _ },
+  le_top := λ K X S hS, set.mem_univ _,
+  ..pretopology.partial_order C }
+
+instance : inhabited (pretopology C) := ⟨⊤⟩
+
+/--
+A pretopology `K` can be completed to a Grothendieck topology `J` by declaring a sieve to be
+`J`-covering if it contains a family in `K`.
+
+See https://stacks.math.columbia.edu/tag/00ZC, or [MM92] Chapter III, Section 2, Equation (2).
+-/
+def to_grothendieck (K : pretopology C) : grothendieck_topology C :=
+{ sieves := λ X S, ∃ R ∈ K X, R ≤ (S : arrows_with_codomain _),
+  top_mem' := λ X, ⟨arrows_with_codomain.singleton_arrow (𝟙 _), K.has_isos _, λ _ _ _, ⟨⟩⟩,
+  pullback_stable' := λ X Y S g,
+  begin
+    rintro ⟨R, hR, RS⟩,
+    refine ⟨_, K.pullbacks g _ hR, _⟩,
+    rw [← sieve.sets_iff_generate, pullback_arrows_comm],
+    apply sieve.pullback_monotone,
+    rwa sieve.gi_generate.gc,
+  end,
+  transitive' :=
+  begin
+    rintro X S ⟨R', hR', RS⟩ R t,
+    choose t₁ t₂ t₃ using t,
+    refine ⟨_, K.transitive _ _ hR' (λ _ f hf, t₂ (RS _ hf)), _⟩,
+    rintro Y _ ⟨Z, g, f, hg, hf, rfl⟩,
+    apply t₃ (RS _ hg) _ hf,
+  end }
+
+lemma mem_to_grothendieck (K : pretopology C) (X S) :
+  S ∈ to_grothendieck C K X ↔ ∃ R ∈ K X, R ≤ (S : arrows_with_codomain X) :=
+iff.rfl
+
+/--
+The largest pretopology generating the given Grothendieck topology.
+
+See [MM92] Chapter III, Section 2, Equations (3,4).
+-/
+def of_grothendieck (J : grothendieck_topology C) : pretopology C :=
+{ coverings := λ X R, sieve.generate R ∈ J X,
+  has_isos := λ X Y f i,
+  begin
+    apply J.covering_of_eq_top,
+    rw [← sieve.id_mem_iff_eq_top],
+    exactI ⟨_, inv f, f, by simp⟩,
+  end,
+  pullbacks := λ X Y f R hR,
+  begin
+    rw [set.mem_def, pullback_arrows_comm],
+    apply J.pullback_stable f hR,
+  end,
+  transitive := λ X S Ti hS hTi,
+  begin
+    apply J.transitive hS,
+    intros Y f,
+    rintros ⟨Z, g, f, hf, rfl⟩,
+    rw sieve.pullback_comp,
+    apply J.pullback_stable g,
+    apply J.superset_covering _ (hTi _ hf),
+    rintro Y g ⟨W, h, g, hg, rfl⟩,
+    exact ⟨_, h, _, ⟨_, _, _, hf, hg, rfl⟩, by simp⟩,
+  end }
+
+/-- We have a galois insertion from pretopologies to Grothendieck topologies. -/
+def insert : galois_insertion (to_grothendieck C) (of_grothendieck C) :=
```

Comment:
```suggestion
def insert : gi (to_grothendieck C) (of_grothendieck C) :=
```

Reply:
Changing `galois_insertion` to `gi` doesn't work for me, did you mean to change the name `insert` to `gi`?

Reply:
@b-mehta Woops, yes. Sorry.